File: Hw04/hw_practice02/code/stitchImages.py
import utils
import numpy as np
from matplotlib import pyplot as plt

from detectCorners import detectCorners
from extractFeatures import extractFeatures
from computeMatches import computeMatches
from mergeImages import mergeImages
from ransac import ransac

# This code is part of:
#
#   CMPSCI 370: Computer Vision, Spring 2016
#   University of Massachusetts, Amherst
#   Instructor: Subhransu Maji
#
#   Homework 4
#
# Read images

im1 = utils.imread('../data/umass_building_left.jpg') #left image
im2 = utils.imread('../data/umass_building_right2.jpg') #right image (change this to right2, right3, ...)

sigma = 1.5
threshold = 0.0005
max_corners = 200
isSimple = False
c1 = detectCorners(im1, isSimple, sigma, threshold)
c2 = detectCorners(im2, isSimple, sigma, threshold)

n1 = min(max_corners, c1.shape[1])
c1 = c1[:, 0:n1]
n2 = min(max_corners, c2.shape[1])
c2 = c2[:, 0:n2]

#Compute feature descriptors
patch_radius = 5
f1 = extractFeatures(im1, c1, patch_radius)
f2 = extractFeatures(im2, c2, patch_radius)

#Compute matches
matches = computeMatches(f1, f2)
utils.showMatches(im1, im2, c1, c2, matches, title='All correspondences')
plt.show()

#Estimate transformation
inliers, transf = ransac(matches, c1, c2)
good_matches = np.zeros_like(matches)-1
good_matches[inliers] = matches[inliers]
utils.showMatches(im1, im2, c1, c2, good_matches, title='Inliers')
plt.show()

#Warp images
stitch_im = mergeImages(im1, im2, transf)
plt.imshow(stitch_im)
plt.title('Stitched Image')
plt.show()

File: Hw04/hw_practice02/code/stitchImages.m
% This code is part of:
%
%   CMPSCI 370: Computer Vision, Spring 2018
%   University of Massachusetts, Amherst
%   Instructor: Subhransu Maji
%
%   Homework 4

% Read images
close all
im1 = imread('../data/umass_building_left.jpg'); % left image
im2 = imread('../data/umass_building_right4.jpg'); % right image (change this to right2, right3,..., right5)

% Detect corners using harris corner detector
sigma = 1.5;
threshold = 0.0005;
maxCorners = 200;
isSimple = false;
c1 = detectCorners(im1, isSimple, sigma, threshold);
c2 = detectCorners(im2, isSimple, sigma, threshold);
n1 = min(maxCorners, size(c1,2));
c1 = c1(:,1:n1);
n2 = min(maxCorners, size(c2,2));
c2 = c2(:,1:n2);

% Compute feature descriptors
patchRadius = 5;
f1 = extractFeatures(im1, c1, patchRadius);
f2 = extractFeatures(im2, c2, patchRadius);


% Compute matches
matches = computeMatches(f1,f2);
showMatches(im1, im2, c1, c2, matches);
title('all matches');

% Estimate transformation
[inliers, transf] = ransac(matches, c1, c2);
goodMatches = zeros(size(matches));
goodMatches(inliers) = matches(inliers);
showMatches(im1, im2, c1, c2, goodMatches);
title('inliers');

% Warp images
stitchIm = mergeImages(im1,im2, transf);
figure;
imshow(stitchIm);
title('stitched image');